### PR TITLE
616: (feat) prevent multiple fidelity bonds with same expiry date

### LINKED ISSUE #616: feat: Prevent creating multiple fidelity bonds with same expiry date
This request stems from a user suggestions. Thank you, btw :pray: 

As the title says, as Jam is tailored towards non-expert users, address reuse should be prevented (as much as possible) when creating fidelity bonds . Since the address of the bond is shown beforehand, if someone really wants to do this - for whatever reason - it can still be done manually or by using the CLI. However, the UI should try to stop this from happening if possible.

Out of scope:
Since Jam does not know any historical, expired (and subsequently moved) fidelity bonds, recreating FB with the same expiry date afterwards _cannot_ be prevented. However, should it be prevented to create multiple FBs in the first place?

### PR DESCRIPTION
**Pull request by barrytra**
This PR fixes #616 
Implementation is done in lockdateForm.js component, that shows a warning message while selecting expiry date whenever a user creates a new fidelity bond.
When is warning message displayed? if there already exists fidelity bonds, and a user create a new one. But he selects an expiry date same as in any of the previous bonds. Then the message is displayed warning the user.

### CODE REVIEW

**Comment by barrytra** [comment]
This is how it works. I have two fidelity bonds already made with expiry date august,2024 and october,2025. There is no warning message when i create a new FB with different expiry date.
Screenshot from 2024-04-13 21-54-26
But as i select date october,2025 for my new FB. It gives a warning like this
Screenshot from 2024-04-16 11-07-15

It works smoothly. The warning vanishes when you set the date to a legal one.
Let me know if you see any issues, especially with the warning message(My English is not that good).
@theborakompanioni

**Comment by editwentyone** [comment]
think about, to remove the text inside () and maybe add this comment instead: https://github.com/joinmarket-webui/jam/issues/616#issuecomment-2058529662

**Comment by theborakompanioni** [comment]
Hey @barrytra! Conecpt ACK!

As discussed, maybe it is a good idea to keep knowledge of Fidelity Bonds out of component `LockdateForm`.
Let me know once you need a review, want a question answered, or need any other help. :pray:

**Comment by barrytra** [comment]
As required, Information of Fidelity bonds has been moved to ```CreateFidelityBond``` file. 
UX works fine as earlier.
Please suggest if any changes required

**Comment by theborakompanioni** [comment]
> As required, Information of Fidelity bonds has been moved to `CreateFidelityBond` file. UX works fine as earlier. Please suggest if any changes required @theborakompanioni

Nice : -)
Looks better and better!

Some notes from my side: 
- The alert is better placed inside the `SelectDate` component, instead of `LockdateForm` - in fact, the LockdateForm component should not need any changes at all.
- Instead of a state variable + function + effect (`useState`/`useEffect`), have you considered using `useMemo`?


e.g.
```
const [lockDateExists, setLockDateExists] = useState(false)

const ifLockTimeExists = (fidelityBonds: Utxos, lockDate: Api.Lockdate): void => {
    setLockDateExists(false)
    if (lockDate)
      // eslint-disable-next-line array-callback-return
      fidelityBonds.map((fidelityBond) => {
        const locktime = fb.utxo.getLocktime(fidelityBond)
        if (locktime === fb.lockdate.toTimestamp(lockDate)) {
          setLockDateExists(true)
        }
      })
  }

  useEffect(() => {
    if (lockDate) ifLockTimeExists(fidelityBonds, lockDate)
  }, [fidelityBonds, lockDate])
```
Can be replaced with something like:
```
const lockDateExists = useMemo(() => {
  return lockDate && fidelityBonds.some(it => fb.utxo.getLocktime(it) === fb.lockdate.toTimestamp(lockDate))
}, [fidelityBonds, lockDate])
```

What do you think?

**Comment by barrytra** [comment]
Ahh yes, that should work and would be much better. And also Alert is shifted to ```createFidelityBond``` component.
I am working on it

**Comment by barrytra** [comment]
Suggested changes are made. Now ```lockdatefrom``` and ```fidelitybondSteps``` components are untouched

**Comment by theborakompanioni** [comment]
> Suggested changes are made. Now `lockdatefrom` and `fidelitybondSteps` components are untouched. This looks fine to me now. what are your suggestions?

Awesome!

Now we just have to adapt the message a little bit and we are ready to merge. :rocket: 

Great work @barrytra!

File: src/i18n/locales/en/translation.json
Comment (theborakompanioni) [inline src/i18n/locales/en/translation.json:452]:
```suggestion
        "warning_fb_with_same_expiry": "<strong>Warning</strong>: A fidelity bond with same expiry date already exists."
```

I think this message is good for now. We can change to a checkbox labelled "I know what I am doing and want to create another fidelity bond" later. What do you think?

Comment (barrytra) [reply src/i18n/locales/en/translation.json:452]: sure we can do that. In such case we need to disable next button until user ticks the checkbox right?
Also should we do it in this PR or do it later?

Comment (theborakompanioni) [reply src/i18n/locales/en/translation.json]: Yes, only in the "en" translation file.

Let's think the checkbox through in a follow-up ticket/PR. The message is already a good improvement! Okay for you?

Comment (barrytra) [reply src/i18n/locales/en/translation.json]: yes.. message looks fine to me.. I commited that change

**Review by theborakompanioni — APPROVED** [review]

**Comment by barrytra** [comment]
I think everything is covered now. Message has been fixed as well. Is there anything else we need to do?

**Comment by theborakompanioni** [comment]
> I think everything is covered now. Message has been fixed as well. Is there anything else we need to do?

Waiting for comments or approval from @editwentyone - then it is ready to be merged! :rocket:

**Pull request merged**
